### PR TITLE
Add code coverage reporting to client build yaml file

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

[AB#15187](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/15187)

Add code coverage reporting to client build yaml file.

### CODE REVIEW

File: packages/loader/container-loader/src/utils.ts [hunk 1]
```diff
@@ -424,3 +424,57 @@ export async function getDocumentAttributes(
 
 	return attributes;
 }
+
+/**
+ * Testing
+ * @internal
+ */
+export async function codeCoverageTesting(
```
Comment (jatgarg) [inline packages/loader/container-loader/src/utils.ts:432]: Will remove this before merging. Just for testing.

File: tools/pipelines/templates/build-npm-package.yml [hunk 2]
```diff
@@ -407,6 +412,41 @@ extends:
...
+                        env:
+                          ADO_API_TOKEN: $(System.AccessToken)
+                          GITHUB_API_TOKEN: $(githubPublicRepoSecret)
+                          TARGET_BRANCH_NAME: '$(targetBranchName)'
+                          ADO_BUILD_ID: $(Build.BuildId)
+                        inputs:
+                          targetType: inline
+                          workingDirectory: '${{ parameters.buildDirectory }}'
+                          script: |
+                            Coverage_Report_Artifact_Name="Code Coverage Report"
```
Comment (alexvy86) [inline tools/pipelines/templates/build-npm-package.yml:428]: Best practice for bash scripts, to make sure any error while executing any line immediately terminates the whole script and thus the ADO step with a failure code.
```suggestion
                            set -eux -o pipefail
                            Coverage_Report_Artifact_Name="Code Coverage Report"
```

File: tools/pipelines/templates/build-npm-package.yml [hunk 3]
```diff
@@ -407,6 +412,41 @@ extends:
...
+                            export CODE_COVERAGE_ANALYSIS_ARTIFACT_NAME_BASELINE=$Coverage_Report_Artifact_Name
+                            export CODE_COVERAGE_ANALYSIS_ARTIFACT_NAME_PR=$Coverage_Report_Artifact_Name
```
Comment (alexvy86) [inline tools/pipelines/templates/build-npm-package.yml:430]: We should refactor `flub report codeCoverage` and get rid of these variables. The command unconditionally adds the build id to the artifact name we pass here, so in reality it assumes that it's going to use the artifact produced by `PublishCodeCoverageResults` whose name we can't control, and that we're passing the correct prefix value here. "Code Coverage Report" is really the only value that would work, so having it be a variable doesn't make sense, IMO. And it'd be hard to make it take the full artifact name for both baseline and PR, because we'd have to find the correct builds here before invoking it. We can keep the flag in the flub command but with a default value and documentation saying "This is the only value that works so you don't need to override it, unless ADO eventually makes changes to the `PublishCodeCoverageResults` task".

File: tools/pipelines/templates/build-npm-package.yml [hunk 4]
```diff
@@ -407,6 +412,41 @@ extends:
...
+                            echo "Code Coverage Artifact Name Prefix for Baseline: $CODE_COVERAGE_ANALYSIS_ARTIFACT_NAME_BASELINE"
+                            echo "Code Coverage Artifact Name Prefix for PR: $CODE_COVERAGE_ANALYSIS_ARTIFACT_NAME_PR"
+                            REPO_NAME=$(echo $BUILD_REPOSITORY_NAME | cut -d'/' -f2)
+                            REPO_OWNER=$(echo $BUILD_REPOSITORY_NAME | cut -d'/' -f1)
+                            export GITHUB_REPOSITORY_NAME=$REPO_NAME
+                            export GITHUB_REPOSITORY_OWNER=$REPO_OWNER
+                            echo "Github Repository Owner: $GITHUB_REPOSITORY_OWNER"
+                            echo "Github Repository Name: $GITHUB_REPOSITORY_NAME"
+                            export GITHUB_PR_NUMBER=$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER
+                            echo "Github PR number: $GITHUB_PR_NUMBER"
```
Comment (alexvy86) [inline tools/pipelines/templates/build-npm-package.yml:440]: Nit: pass this as `GITHUB_PR_NUMBER: $(System.PullRequest.PullRequestNumber)` from the `env:` property of the task? Since it doesn't require any special computation in the script.

I'd do the same for `ADO_CI_BUILD_DEFINITION_ID_BASELINE` and `ADO_CI_BUILD_DEFINITION_ID_PR` for the same reason. The less we have in the content of the script, the better, IMO.

File: tools/pipelines/templates/include-test-task.yml [hunk 5]
```diff
@@ -15,8 +15,8 @@ parameters:
   default: false
 
 steps:
-# Test - With coverage
-- ${{ if and(parameters.testCoverage, startsWith(parameters.taskTestStep, 'ci:test')) }}:
+# Test - With coverage - Don't run with coverage if the task is a PR and the task is a tinylicious service and stress test.
+- ${{ if and(parameters.testCoverage, startsWith(parameters.taskTestStep, 'ci:test'), or(ne(variables['Build.Reason'], 'PullRequest'), and(not(contains(parameters.taskTestStep, 'realsvc:tinylicious')), not(contains(parameters.taskTestStep, 'stress:tinylicious'))))) }}:
```
Comment (alexvy86) [inline tools/pipelines/templates/include-test-task.yml:19]: It's pretty hard to parse otherwise. I think we also need to say _why_ we're skipping coverage for those tests in PRs (to match what we do in the bundle analysis pipeline, IIRC based on what we discussed, right?).
```suggestion
- ${{ if and(
	parameters.testCoverage,
	startsWith(parameters.taskTestStep, 'ci:test'),
	or(
		ne(variables['Build.Reason'], 'PullRequest'),
		and(
			not(contains(parameters.taskTestStep, 'realsvc:tinylicious')),
			not(contains(parameters.taskTestStep, 'stress:tinylicious'))
		)
	)
) }}:
```

Comment (jatgarg) [reply tools/pipelines/templates/include-test-task.yml]: Added the comment, but indentation wise, it needs to be in 1 line otherwise yaml complain about the structure and does not recognize it as a condition.

Comment (alexvy86) [reply tools/pipelines/templates/include-test-task.yml]: Hmm did it complain both with tabs and spaces? We have some [working examples](https://github.com/microsoft/FluidFramework/blob/fcd15e0cb38d037b6f3dfbd4a36f7850190a16d1/tools/pipelines/build-docs.yml#L73-L76) of expressions spread across more than 1 line, but I couldn't find one for an `if` expression. If it doesn't work, oh well.

File: .vscode/launch.json [hunk 6]
```diff
@@ -20,6 +20,27 @@
...
+				// Your ADO PAT (Personal Access Token) User settings > Personal Access Tokens > New Token (check read permissions, likely Build, Release, Packaging, and Test Management)
+				"ADO_API_TOKEN": "",
+				"GITHUB_API_TOKEN": "",
+				"GITHUB_PR_NUMBER": "",
+				"ADO_BUILD_ID": "",
+				"ADO_CI_BUILD_DEFINITION_ID_BASELINE": "48",
+				"ADO_CI_BUILD_DEFINITION_ID_PR": "11",
+				"TARGET_BRANCH_NAME": "main",
+				"GITHUB_REPOSITORY_OWNER": "microsoft",
+				"GITHUB_REPOSITORY_NAME": "FluidFramework",
```
Comment (tylerbutler) [inline .vscode/launch.json:41]: While you can use the env variables for this, for local testing it's typically easier to pass at least some of them as flags instead of env variables.

File: package.json [hunk 7]
```diff
@@ -53,6 +53,7 @@
 		"clean:nyc": "rimraf --glob \"nyc/**\"",
 		"clean:r11s": "fluid-build -g server --task clean",
 		"clean:root": "rimraf --glob \"*.done.build.log\"",
+		"code-coverage-analysis:report": "flub report codeCoverage",
```
Comment (tylerbutler) [inline package.json:56]:
```suggestion
		"codeCoverage:report": "flub report codeCoverage",
```

File: tools/pipelines/templates/build-npm-package.yml [hunk 8]
```diff
@@ -407,6 +412,37 @@ extends:
...
+                          GITHUB_PR_NUMBER: $(System.PullRequest.PullRequestNumber)
+                          ADO_CI_BUILD_DEFINITION_ID_BASELINE: 48
+                          ADO_CI_BUILD_DEFINITION_ID_PR: 11
+                        inputs:
+                          targetType: inline
+                          workingDirectory: '${{ parameters.buildDirectory }}'
+                          script: |
+                            set -eux -o pipefail
+                            REPO_NAME=$(echo $BUILD_REPOSITORY_NAME | cut -d'/' -f2)
+                            REPO_OWNER=$(echo $BUILD_REPOSITORY_NAME | cut -d'/' -f1)
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:433]: No using bash!

File: tools/pipelines/templates/build-npm-package.yml [hunk 9]
```diff
@@ -407,6 +412,37 @@ extends:
...
+                            export GITHUB_REPOSITORY_NAME=$REPO_NAME
+                            export GITHUB_REPOSITORY_OWNER=$REPO_OWNER
+                            echo "Github Repository Owner: $GITHUB_REPOSITORY_OWNER"
+                            echo "Github Repository Name: $GITHUB_REPOSITORY_NAME"
+                            echo "Github PR number: $GITHUB_PR_NUMBER"
+                            echo "ADO Build Number: $ADO_BUILD_ID"
+                            echo "Target Branch Name: $TARGET_BRANCH_NAME"
+                            echo "ADO CI BUILD_DEFINITION_ID for baseline: $ADO_CI_BUILD_DEFINITION_ID_BASELINE"
+                            echo "ADO CI BUILD_DEFINITION_ID for PR: $ADO_CI_BUILD_DEFINITION_ID_PR"
+                            echo "Running code coverage comparison"
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:443]: Seems like all this logging can be done in the command itself instead of the pipeline.

File: tools/pipelines/templates/build-npm-package.yml (hunk 8 above)
Comment (tylerbutler) [reply tools/pipelines/templates/build-npm-package.yml:433]: Instead of splitting this value here, have the command accept a single flag with the combined value and split in the command if you need to.

File: tools/pipelines/templates/include-test-task.yml [hunk 10]
```diff
@@ -15,8 +15,8 @@ parameters:
   default: false
 
 steps:
-# Test - With coverage
-- ${{ if and(parameters.testCoverage, startsWith(parameters.taskTestStep, 'ci:test')) }}:
+# Test - With coverage - Don't run with coverage if the task is a PR and the task is a tinylicious service and stress test. We don't include coverage for these tests in our baseline coverage stats, so we don't need to run them with coverage here too.
+- ${{ if and(parameters.testCoverage, startsWith(parameters.taskTestStep, 'ci:test'), or(ne(variables['Build.Reason'], 'PullRequest'), and(not(contains(parameters.taskTestStep, 'realsvc:tinylicious')), not(contains(parameters.taskTestStep, 'stress:tinylicious')))))}}:
```
Comment (tylerbutler) [inline tools/pipelines/templates/include-test-task.yml:19]: You can probably make this a variable which will let you use better formatting, and then the if check here can just look at the variable. E.g.

https://github.com/microsoft/FluidFramework/blob/b053eb67bd6791349e2ac277697a8b2628f13864/tools/pipelines/templates/include-vars.yml#L28-L43

File: package.json (hunk 7 above)
Comment (tylerbutler) [reply package.json:56]: Is there a "collect" step like there is for bundle size?

File: tools/pipelines/templates/build-npm-package.yml [hunk 11]
```diff
@@ -407,6 +412,37 @@ extends:
...
+                            npm run code-coverage-analysis:report
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:444]: You could also call flub directly here, which would use the in-repo version of the report:codeCoverage command instead of the installed version. Might be a lot faster for testing and bug fixing. You don't need a build-tools release in that case.

Comment (tylerbutler) [reply tools/pipelines/templates/build-npm-package.yml:444]: That is:

```suggestion
                            flub report codeCoverage
```

File: tools/pipelines/templates/include-test-task.yml (hunk 10 above)
Comment (alexvy86) [reply tools/pipelines/templates/include-test-task.yml:19]: Note: template yml files cannot declare variables, so I think what we'd need to do to implement this is to put the whole complex condition in the "caller file" where it passes the `testCoverage` parameter, and here keep a simple `${{ if parameters.testCoverage }}`. I like the idea though.

File: tools/pipelines/templates/build-npm-package.yml (hunk 11 above)
Comment (alexvy86) [reply tools/pipelines/templates/build-npm-package.yml:444]: If we make the change, then I'd say remove the npm script altogether, so people are not misguided into using it and thinking that it should have the same behavior as what happens in the pipeline (it would be using the build-cli/flub installed by the package, not the in-repo version).

File: .vscode/launch.json (hunk 6 above)
Comment (jatgarg) [reply .vscode/launch.json:41]: So, the oclif will collect from both whether we send them as flags or env variables?

File: tools/pipelines/templates/build-npm-package.yml (hunk 8 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:433]: What is the issue with bash and if there is some issue, I should be able to use something else?
I still want to split it here rather than in the tool, because both are worked with separately in the git APIs, rather than together. So, I prefer to split here.

File: package.json (hunk 7 above)
Comment (jatgarg) [reply package.json:56]: The tests create the coverage stats when run with coverage command and the Publish step/task in yml, publish those into artifacts.

File: tools/pipelines/templates/build-npm-package.yml (hunk 11 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:444]: @alexvy86 You mean if we do this, then remove the codeCovearge:report command from the root package.json?

File: tools/pipelines/templates/build-npm-package.yml (hunk 8 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:433]: Is it about the combability with windows?

File: tools/pipelines/templates/build-npm-package.yml (hunk 11 above)
Comment (alexvy86) [reply tools/pipelines/templates/build-npm-package.yml]: Yeah. It's not the end of the world to have it there, but it won't use the same flub version that the pipeline will be using here (the one installed from npm in package.json, the latest version in the repo in the pipeline). We just need to be aware of that.

File: .vscode/launch.json (hunk 6 above)
Comment (jatgarg) [reply .vscode/launch.json:41]: And easier in what sense? Since we are passing them like this in yml pipeline, we used it like this here.

File: tools/pipelines/templates/build-npm-package.yml (hunk 8 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:442]: .

Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml]: Removed it

File: tools/pipelines/templates/build-npm-package.yml (hunk 9 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml]: Only a few lines left now.  Without error, we only log verbose logs which does not get printed in the pipeline in ADO. I think it is ok here.

File: tools/pipelines/templates/build-npm-package.yml [hunk 12]
```diff
@@ -407,6 +421,33 @@ extends:
...
                           sourceFolder: '${{ parameters.buildDirectory }}/nyc/report'
                           targetFolder: $(Build.ArtifactStagingDirectory)/codeCoverageAnalysis
+                      - task: Bash@3
+                        displayName: Report Code Coverage Comparison
+                        condition: and(succeededOrFailed(), eq('${{ parameters.reportCodeCoverageComparison }}', true), eq(variables['ReportDirExists'], 'true'), eq(variables['Build.Reason'], 'PullRequest'), eq(variables['System.PullRequest.TargetBranch'], 'main'))
+                        continueOnError: false
+                        env:
+                          ADO_API_TOKEN: $(System.AccessToken)
+                          GITHUB_API_TOKEN: $(githubPublicRepoSecret)
+                          TARGET_BRANCH_NAME: '$(targetBranchName)'
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:431]: Should all of these be in quotes?

File: tools/pipelines/templates/build-npm-package.yml [hunk 13]
```diff
@@ -407,6 +421,33 @@ extends:
                         inputs:
                           sourceFolder: '${{ parameters.buildDirectory }}/nyc/report'
                           targetFolder: $(Build.ArtifactStagingDirectory)/codeCoverageAnalysis
+                      - task: Bash@3
+                        displayName: Report Code Coverage Comparison
+                        condition: and(succeededOrFailed(), eq('${{ parameters.reportCodeCoverageComparison }}', true), eq(variables['ReportDirExists'], 'true'), eq(variables['Build.Reason'], 'PullRequest'), eq(variables['System.PullRequest.TargetBranch'], 'main'))
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:426]: Are the extra conditions still needed with the reportCodeCoverageComparison parameter?

File: tools/pipelines/templates/build-npm-package.yml [hunk 14]
```diff
@@ -576,10 +617,11 @@ extends:
                       sbomEnabled: false
                       publishLocation: pipeline
 
+                # Publish thec Code Coverage  artifact in this format as it is easier for devs to find coverage html for each file.
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:620]:
```suggestion
                # Publish the Code Coverage  artifact in this format as it is easier for devs to find coverage html for each file.
```

File: tools/pipelines/templates/build-npm-package.yml (hunk 13 above)
Comment (tylerbutler) [reply tools/pipelines/templates/build-npm-package.yml:426]: `eq(variables['Build.Reason'], 'PullRequest')`

that one seems to be duplicated.

`eq(variables['System.PullRequest.TargetBranch'], 'main')`

Could that be integrated into the parameter instead? In build-client.yml?

File: tools/pipelines/templates/build-npm-package.yml (hunk 12 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:431]: Works just fine like this.

File: tools/pipelines/templates/build-npm-package.yml (hunk 13 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml:426]: Yeah, the first one is duplicated. Removed that.

File: tools/pipelines/build-client.yml [hunk 15]
```diff
@@ -188,7 +188,8 @@ extends:
     interdependencyRange: ${{ parameters.interdependencyRange }}
     packageManagerInstallCommand: 'pnpm i --frozen-lockfile'
     packageManager: pnpm
-    testCoverage: ${{ variables.testCoverage }}
+    testCoverage: true
```
Comment (alexvy86) [inline tools/pipelines/build-client.yml:191]: Tangent to this PR, no need to take action now: @scarlettjlee @agarwal-navin this is a point to investigate when optimizing the internal pipeline runtime. We want the tests to run with code coverage in the public project so we can calculate the coverage deltas for PRs, but if we don't leverage that data in the internal project we could update this condition to `testCoverage: ${{ eq(variables['System.TeamProject'], 'public' ) }}` if there are significant runtime wins from skipping it in internal.

**Review by alexvy86 — APPROVED** [review]
A few last comments from me. I think adding the validation in the flub command is worth doing before we merge (I know we then have to publish and consume a new release of build-tools for the logic to actually be used).

I'd say still double check with @tylerbutler  if he has any more blocking feedback.

File: tools/pipelines/templates/build-npm-package.yml (hunk 13 above)
Comment (jatgarg) [reply tools/pipelines/templates/build-npm-package.yml]: Added this in reportCodeCoverageComparison itself.

File: tools/pipelines/templates/build-npm-package.yml (hunk 9 above)
Comment (tylerbutler) [reply tools/pipelines/templates/build-npm-package.yml]: > Without error, we only log verbose logs which does not get printed in the pipeline in ADO.

No need to change this PR, but note that we can run `flub --verbose` in CI if we want the verbose logs. We do that in other places where the verbose logging is very helpful and not too spammy.

File: tools/pipelines/templates/build-npm-package.yml (hunk 12 above)
Comment (tylerbutler) [reply tools/pipelines/templates/build-npm-package.yml:433]: @alexvy86 might know more, but I think that if the variable has any characters like `:` in it then the YAML parsing will fail. YAML treats an unquoted string as a string but quotes are required if it contains "special characters" but I don't know the full list. This is probably fine, but the API tokens in particular can sometimes have odd characters in them. We can wait to see if it's a real problem.

File: tools/pipelines/templates/build-npm-package.yml [hunk 16]
```diff
@@ -576,10 +619,11 @@ extends:
                       sbomEnabled: false
                       publishLocation: pipeline
 
+                # Publish the Code Coverage  artifact in this format as it is easier for devs to find coverage html for each file.
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-npm-package.yml:622]: nit

```suggestion
                # Publish the Code Coverage artifact in this format as it is easier for devs to find coverage html for each file.
```

**Review by tylerbutler — APPROVED** [review]
Just nits at this point. Thanks for diligently working through all the feedback!

File: tools/pipelines/templates/build-npm-package.yml (hunk 12 above)
Comment (alexvy86) [reply tools/pipelines/templates/build-npm-package.yml:433]: I think in this case it wouldn't matter. If we were using template expressions (`${{ }}`) then maybe, because those get replaced during the "compilation" (syntax checking, etc) phase of the pipeline, where weird characters could indeed break the YAML structure. `$()` variables get replaced at runtime, I think right when the task is about to execute, so the concerns would be more on the side of "will bash interpret this correctly if it has spaces or weird characters?" and most of the time I think we're fine.

Comment (alexvy86) [reply tools/pipelines/templates/build-npm-package.yml:433]: Oh, I see we added the quotes. I believe ADO strips the quotes before setting the env variable values, so we end up with `token` and not `"token"`, so we should be fine.

File: tools/pipelines/templates/build-npm-package.yml [hunk 17]
```diff
@@ -425,7 +425,7 @@ extends:
                           targetFolder: $(Build.ArtifactStagingDirectory)/codeCoverageAnalysis
                       - task: Bash@3
                         displayName: Report Code Coverage Comparison
-                        condition: and(succeededOrFailed(), eq('${{ parameters.reportCodeCoverageComparison }}', true), eq(variables['ReportDirExists'], 'true'))
+                        condition: and(succeededOrFailed(), eq(parameters.reportCodeCoverageComparison, true), eq(variables['ReportDirExists'], 'true'))
```
Comment (alexvy86) [inline tools/pipelines/templates/build-npm-package.yml]: What was the problem? I don't think this is going to work, all access to `paramaters.*` needs to be inside template expressions `${{}}` AFAIK.

Comment (alexvy86) [reply tools/pipelines/templates/build-npm-package.yml]: The single quotes around `${{}}` were probably unnecessary though.
